src/ddmd/root/filename.d

```diff
@@ -721,3 +724,111 @@ nothrow:
+      https://msdn.microsoft.com/en-us/library/windows/desktop/aa363855(v=vs.85).aspx
+    */
+    private int _mkdir(const(char)* path) nothrow
+    {
+        import core.sys.windows.winerror: ERROR_ALREADY_EXISTS;
+        import core.stdc.errno: errno, EEXIST;
+
+        const extendedPath = path.toExtendedLengthPath();
+
+        // the CreateDirectoryW function takes a pointer to a structure,
+        // but since we're not interested in using it it's declared void*
+        // and we pass null for the default behaviour.
+        void* securityAttributes;
+        const createRet = () @trusted
+        {
+            return CreateDirectoryW(&extendedPath[0], securityAttributes);
+        }();
+        const lastError = GetLastError();
+
+        // Preserve compatibility with mkdir since the calling code expects
+        // errno to be set.
+        if (createRet == 0 && lastError == ERROR_ALREADY_EXISTS)
+            errno(EEXIST);
+
+        // different conventions for CreateDirectory and mkdir
+        return createRet == 0 ? 1 : 0;
+    }
+
+    // Converts an UTF8 null-terminated string to an array of wchar that's null
+    // terminated so it can be passed to Win32 APIs.
+    private wchar[] toWStringz(const(char*) str) pure nothrow {
+        import core.stdc.string: strlen;
+
+        const length8 = strlen(str);
+
+        // Conservative estimate that should be enough for any UTF8 string
+        // according to https://tools.ietf.org/html/rfc3629.
+        // The +1 is for the null terminator.
+        const length16 = length8 * 4 + 1;
+
+        auto wstr = new wchar[length16];
+
+        // Using i, wchar c in the foreach doesn't work since then i would
+        // be tied to the length of the UTF8 sequence.
+        int wstrLen;
+        try
+            foreach(wchar c; str[0 .. length8]) wstr[wstrLen++] = c;
+        catch(Exception)
+            return null;
+
+        wstr[wstrLen] = 0; // null-terminate it
+
+        return wstr[0 .. wstrLen];
+    }
+
+    // Converts a path to one suitable to be passed to Win32 API
+    // functions that can deal with paths longer than 248
+    // characters. For more information:
+    // https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx
+    package wchar[] toExtendedLengthPath(const(char*) path) pure nothrow {
+        // Since the unicode Win32 APIs use UTF16, we need to convert
+        // from UTF8 to UTF16 first.
+        const wpath = path.toWStringz();
+
+        // Now wpath is the UTF16 version of path, but to be able to use
+        // extended paths, we need to prefix with `\\?\` and the absolute
+        // path.
+
+        static immutable prefix = `\\?\`w; // see comment above
+
+        // We don't know how large the absolute file path will be We
+        // need to tell GetFullPathName the size of the buffer we're
+        // passing in so we defensively allocate 4 times the size of
+        // the string we have currently.
```

Comment:
If you call _wfullpath with a null pointer as a pointer to the buffer to store the absolute path then it calls malloc for you, and you just need to free the result.  Presuming that the behaviour of _wfullpath is identical in this respect to _fullpath, which I think is true.

https://docs.microsoft.com/en-gb/cpp/c-runtime-library/reference/fullpath-wfullpath
"The _fullpath function expands the relative path name in relPath to its fully qualified or absolute path and stores this name in absPath. If absPath is NULL, malloc is used to allocate a buffer of sufficient length to hold the path name. It is the responsibility of the caller to free this buffer. "

I might have misunderstood what this does - possibly it's not what we want.  (I'm not on a Windows machine right now).  But I'm not a Windows guy and yet it didn't take long to find the function (and as so often the case, when you do it the right way it solves other problems you didn't think were related).  I just figured the function must exist, and googled for it.  I had the advantage of knowing I didn't know Windows API at all, which meant I didn't make assumptions - so if we could move a bit towards operating that way more it would help, I think.

Reply:
Operating that way meaning identifying what the problem is precisely, and checking to see if there's a better answer.